src/iterative_ensemble_smoother/esmda.py

```diff
@@ -0,0 +1,255 @@
+"""
+Ensemble Smoother with Multiple Data Assimilation (ES-MDA)
+----------------------------------------------------------
+
+Imlementation of the 2013 paper "Ensemble smoother with multiple data assimilation"
+
+References
+----------
+
+Emerick, A.A., Reynolds, A.C. History matching time-lapse seismic data using
+the ensemble Kalman filter with multiple data assimilations.
+Comput Geosci 16, 639–659 (2012). https://doi.org/10.1007/s10596-012-9275-5
+
+Alexandre A. Emerick, Albert C. Reynolds.
+Ensemble smoother with multiple data assimilation.
+Computers & Geosciences, Volume 55, 2013, Pages 3-15, ISSN 0098-3004,
+https://doi.org/10.1016/j.cageo.2012.03.011.
+
+https://gitlab.com/antoinecollet5/pyesmda
+
+https://helper.ipam.ucla.edu/publications/oilws3/oilws3_14147.pdf
+
+"""
+
+import numbers
+from typing import Optional, Union
+
+import numpy as np
+import numpy.typing as npt
+import scipy as sp  # type: ignore
+
+from iterative_ensemble_smoother.esmda_inversion import (
+    inversion_exact_cholesky,
+    inversion_subspace,
+    normalize_alpha,
+)
+
+
+class ESMDA:
+    # Available inversion methods. The inversion methods all compute
+    # C_MD @ (C_DD + alpha * C_D)^(-1)  @ (D - Y)
+    _inversion_methods = {
+        "exact": inversion_exact_cholesky,
+        "subspace": inversion_subspace,
+    }
+
+    def __init__(
+        self,
+        C_D: npt.NDArray[np.double],
+        observations: npt.NDArray[np.double],
+        alpha: Union[int, npt.NDArray[np.double]] = 5,
+        seed: Optional[int] = None,
```

Comment:
Nitpick, but I guess this should be typed as a union of int and generator.

Reply:
True! fixed!